lib/graphql/analysis/ast/query_complexity.rb

```diff
@@ -145,35 +149,38 @@ def merged_max_complexity(query, inner_selections)
 
           # Add up the total cost for each unique field name's coalesced selections
           unique_field_keys.each_key.reduce(0) do |total, field_key|
-            composite_scopes = nil
-            field_cost = 0
-
-            # Collect composite selection scopes for further aggregation,
-            # leaf selections report their costs directly.
-            inner_selections.each do |inner_selection|
-              child_scope = inner_selection[field_key]
-              next unless child_scope
-
-              # Empty child scopes are leaf nodes with zero child complexity.
-              if child_scope.empty?
-                field_cost = child_scope.own_complexity(0)
-                field_complexity(child_scope, max_complexity: field_cost, child_complexity: nil)
+            # Collect all child scopes for this field key;
+            # all keys come with at least one scope.
+            child_scopes = inner_selections.filter_map { _1[field_key] }
+
+            # Compute maximum possible cost of child selections;
+            # composites merge their maximums, while leaf scopes are always zero.
+            # FieldsWillMerge validation assures all scopes are uniformly one or the other.
```

Comment:
With `FieldsWillMerge` getting fixed, this should now be consistent about always having scopes of the same kind. Either way, it's written to be fault-tolerant in the event that it's run without first validating.